File: scripts/run_validation_semiauto.py
#!/usr/bin/env python3

"""Wrapper semiautomatico para ejecutar validation_1 con datos custom.

- Respaldar los CSV fijos de data/for_validation
- Copiar los archivos provistos a los nombres esperados por validation_1
- Ejecutar run_validation_all.py
- Restaurar los archivos originales al final
"""

from __future__ import annotations

import os
import shutil
import subprocess
import sys
from pathlib import Path


def main() -> int:
  if len(sys.argv) != 4:
    print(
      'Uso: python3 scripts/run_validation_semiauto.py '
      '"ruta/homogeneity.csv" "ruta/stability.csv" "ruta/summary.csv"',
      file=sys.stderr,
    )
    return 2

  project_root = Path.cwd().resolve()
  validation_dir = project_root / "validation_1"
  target_dir = project_root / "data" / "for_validation"

  source_files = {
    "homogeneity": Path(sys.argv[1]).expanduser().resolve(),
    "stability": Path(sys.argv[2]).expanduser().resolve(),
    "summary": Path(sys.argv[3]).expanduser().resolve(),
  }
  expected_files = {
    "homogeneity": target_dir / "homogeneity_n4.csv",
    "stability": target_dir / "stability_n4.csv",
    "summary": target_dir / "summary_n4.csv",
  }
  backup_files = {
    name: Path(str(path) + ".backup_semiauto")
    for name, path in expected_files.items()
  }

  for name, path in source_files.items():
    if not path.is_file():
      raise FileNotFoundError(f"No existe el archivo de entrada: {path}")

  def copy_file(src: Path, dst: Path) -> None:
    dst.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(src, dst)

  try:
    for name, dst in expected_files.items():
      if dst.exists():
        shutil.copy2(dst, backup_files[name])

    for name, dst in expected_files.items():
      copy_file(source_files[name], dst)

    subprocess.run(
      [
        "python3",
        "-c",
        (
          "import os, sys; "
          "sys.path.insert(0, os.getcwd()); "
          "from stage_01_robust_stats import run_stage_01_robust_stats; "
          "from stage_02_homogeneity import run_stage_02; "
          "from stage_03_stability import run_stage_03; "
          "from stage_04_uncertainty_chain import run_stage_04; "
          "from stage_05_scores import run_stage_05; "
          "import stage_05_scores as s5; s5.DATA_PT_DATA = '../data/pt_data_n13.csv'; "
          "run_stage_01_robust_stats(); "
          "run_stage_02(); "
          "run_stage_03(); "
          "run_stage_04(); "
          "run_stage_05()"
        ),
      ],
      cwd=validation_dir,
      check=True,
    )
    print("\nValidación semiautomática completada.")
    return 0
  finally:
    for name, dst in expected_files.items():
      bak = backup_files[name]
      if bak.exists():
        shutil.copy2(bak, dst)
        bak.unlink()


if __name__ == "__main__":
  raise SystemExit(main())

File: scripts/run_validation_semiauto.R
#!/usr/bin/env Rscript

# ===================================================================
# Wrapper semiautomatico para ejecutar validation_1 con datos custom
#
# - Respaldar los archivos fijos de validation_1/data/for_validation
# - Copiar los CSV provistos por argumento a los nombres esperados
# - Ejecutar run_validation_all.R
# - Restaurar los archivos originales al final
# ===================================================================

args <- commandArgs(trailingOnly = TRUE)

if (length(args) != 3) {
  stop(
    "Uso: Rscript scripts/run_validation_semiauto.R ",
    "\"ruta/homogeneity.csv\" \"ruta/stability.csv\" \"ruta/summary.csv\""
  )
}

homogeneity_src <- normalizePath(args[[1]], mustWork = TRUE)
stability_src <- normalizePath(args[[2]], mustWork = TRUE)
summary_src <- normalizePath(args[[3]], mustWork = TRUE)

project_root <- normalizePath(getwd(), mustWork = TRUE)
validation_dir <- file.path(project_root, "validation_1")
target_dir <- file.path(project_root, "data", "for_validation")

expected_files <- c(
  homogeneity = file.path(target_dir, "homogeneity_n4.csv"),
  stability = file.path(target_dir, "stability_n4.csv"),
  summary = file.path(target_dir, "summary_n4.csv")
)
source_files <- c(
  homogeneity = homogeneity_src,
  stability = stability_src,
  summary = summary_src
)
backup_files <- paste0(expected_files, ".backup_semiauto")
names(backup_files) <- names(expected_files)

copy_file <- function(src, dst) {
  dir.create(dirname(dst), recursive = TRUE, showWarnings = FALSE)
  ok <- file.copy(src, dst, overwrite = TRUE)
  if (!isTRUE(ok)) {
    stop("No se pudo copiar: ", src, " -> ", dst)
  }
}

restore_files <- function() {
  for (name in names(expected_files)) {
    dst <- expected_files[[name]]
    bak <- backup_files[[name]]
    if (file.exists(bak)) {
      file.copy(bak, dst, overwrite = TRUE)
      file.remove(bak)
    }
  }
}

for (name in names(expected_files)) {
  dst <- expected_files[[name]]
  bak <- backup_files[[name]]
  if (file.exists(dst)) {
    file.copy(dst, bak, overwrite = TRUE)
  }
}

on.exit(restore_files(), add = TRUE)

for (name in names(expected_files)) {
  copy_file(source_files[[name]], expected_files[[name]])
}

setwd(validation_dir)
source("stage_01_robust_stats.R")
run_stage_01_robust_stats()

source("stage_02_homogeneity.R")
run_stage_02()

source("stage_03_stability.R")
run_stage_03()

source("stage_04_uncertainty_chain.R")
run_stage_04()

source("stage_05_scores.R")
DATA_PT_DATA <- "../data/pt_data_n13.csv"
run_stage_05()

cat("\nValidación semiautomática completada.\n")
